src/pip/_internal/req/__init__.py

```
@@ -60,39 +70,117 @@ def install_given_reqs(
+                partial(_single_install, install_args, suppress_exception=True),
+                should_parallel_values
+            )
+            # collect back results
+            parallel_install_results = dict(
+                zip(should_parallel_indexes, executed_parallel_reqs))
+        else:
+            parallel_install_results = {}
+
+        # check the results from the parallel installation,
+        # and fill-in missing installations or raise exception
+        for i, req in enumerate(requirements):
+
+            # select the install result from the parallel installation
+            # or install serially now
             try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
+                install_result = parallel_install_results[i]
+                if isinstance(install_result, InstallationResult):
+                    logger.debug(
+                        'Successfully installed %s in parallel', req.name)
+            except KeyError:
+                install_result = _single_install(
+                    install_args, req, suppress_exception=False)
+                logger.debug('Successfully installed %s serially', req.name)
```

This logs a successful install but just after, it may throw an exception: `isinstance(install_result, BaseException)` ?

---

@sbidoul when calling `_single_install`  with `suppress_exception=False` it will not suppress the exception but raise it. This means you will not get `isinstance(install_result, BaseException)` in this section, hence if we passed the `_single_install` call, the installation was successful.